File: sdks/swift/Tests/VouchCoreTests/VouchFleetTests.swift
import XCTest

@testable import VouchCore

final class VouchFleetTests: XCTestCase {
    private func trustedRoot(_ root: VouchAgent) -> [String: Data] {
        [root.did: root.publicKey]
    }

    private func signDeviceAction(_ device: VouchAgent, grant: String, resource: String) throws -> String {
        let grantView = VouchCredentials.Credential(grant)
        return try device.sign(action: grantView.action!, target: grantView.target!, resource: resource)
    }

    func testEnrollAndVerifyChain() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create() // did:key

        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")

        let result = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: trustedRoot(root))
        XCTAssertTrue(result.ok, result.reason ?? "")
        XCTAssertEqual(result.rootDid, root.did)
        XCTAssertEqual(result.leaf?.issuer, device.did)
        XCTAssertEqual(result.leaf?.resource, "https://api.bank/invoices/42")
    }

    func testUntrustedRootRejected() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")

        let result = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: [:])
        XCTAssertFalse(result.ok)
    }

    func testWrongDeviceIssuerRejected() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let impostor = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(impostor, grant: grant, resource: "https://api.bank/invoices/42")

        let result = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: trustedRoot(root))
        XCTAssertFalse(result.ok)
    }

    func testTamperedActionRejected() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")
            .replacingOccurrences(of: "invoices/42", with: "invoices/evil")

        let result = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: trustedRoot(root))
        XCTAssertFalse(result.ok)
    }

    func testLeafIntentPolicy() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")
        let roots = trustedRoot(root)

        let ok = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: roots, requireAction: "charge")
        XCTAssertTrue(ok.ok, ok.reason ?? "")

        let bad = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: roots, requireAction: "refund")
        XCTAssertFalse(bad.ok)
    }

    func testRevokedDeviceRejected() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")
        let roots = trustedRoot(root)

        let before = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: roots)
        XCTAssertTrue(before.ok, before.reason ?? "")

        let after = VouchFleet.verifyDelegatedChain(
            [grant, action], trustedRoots: roots, revoked: { $0 == device.did })
        XCTAssertFalse(after.ok)
    }

    func testDeviceRegistryTracksRevocation() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(
            root: root, deviceDid: device.did, action: "charge", target: "api.bank", resource: "https://api.bank/invoices")
        let action = try signDeviceAction(device, grant: grant, resource: "https://api.bank/invoices/42")
        let roots = trustedRoot(root)

        let registry = VouchFleet.DeviceRegistry()
        registry.enroll(device.did, grant: grant)
        XCTAssertEqual(registry.activeDevices().count, 1)

        let before = VouchFleet.verifyDelegatedChain(
            [grant, action], trustedRoots: roots, revoked: registry.isRevoked)
        XCTAssertTrue(before.ok, before.reason ?? "")

        registry.revoke(device.did)
        XCTAssertEqual(registry.activeDevices().count, 0)
        let after = VouchFleet.verifyDelegatedChain(
            [grant, action], trustedRoots: roots, revoked: registry.isRevoked)
        XCTAssertFalse(after.ok)
    }

    func testDidKeyLinkResolvesWithoutTrustMap() throws {
        let root = try VouchAgent.create(domain: "root.example")
        let device = try VouchAgent.create()
        let grant = try VouchFleet.enrollDevice(root: root, deviceDid: device.did, action: "read", target: "t", resource: "https://x/y")
        let action = try signDeviceAction(device, grant: grant, resource: "https://x/y/z")

        let result = VouchFleet.verifyDelegatedChain([grant, action], trustedRoots: trustedRoot(root))
        XCTAssertTrue(result.ok, result.reason ?? "")
    }
}

File: sdks/dotnet/VouchCore.Tests/VouchFleetTests.cs
using System.Collections.Generic;
using VouchProtocol.Core;
using Xunit;

namespace VouchCore.Tests;

/// <summary>Tests for VouchFleet (cross-device identity and delegation).</summary>
public class VouchFleetTests
{
    private static Dictionary<string, string> TrustedRoot(VouchAgent root)
        => new() { [root.Did] = root.PublicKeyB64 };

    private static string SignDeviceAction(VouchAgent device, string grant, string resource)
    {
        var grantView = new VouchCredentials.Credential(grant);
        return device.Sign(grantView.Action!, grantView.Target!, resource);
    }

    [Fact]
    public void EnrollAndVerifyChain()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create(); // did:key

        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42");

        var result = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, TrustedRoot(root));
        Assert.True(result.Ok, result.Reason);
        Assert.Equal(root.Did, result.RootDid);
        Assert.Equal(device.Did, result.Leaf!.Issuer);
        Assert.Equal("https://api.bank/invoices/42", result.Leaf!.Resource);
    }

    [Fact]
    public void UntrustedRootRejected()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42");

        var result = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, new Dictionary<string, string>());
        Assert.False(result.Ok);
    }

    [Fact]
    public void WrongDeviceIssuerRejected()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        VouchAgent impostor = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(impostor, grant, "https://api.bank/invoices/42");

        var result = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, TrustedRoot(root));
        Assert.False(result.Ok);
    }

    [Fact]
    public void TamperedActionRejected()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42")
            .Replace("invoices/42", "invoices/evil");

        var result = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, TrustedRoot(root));
        Assert.False(result.Ok);
    }

    [Fact]
    public void LeafIntentPolicy()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42");
        var roots = TrustedRoot(root);

        var ok = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots, requireAction: "charge");
        Assert.True(ok.Ok, ok.Reason);

        var bad = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots, requireAction: "refund");
        Assert.False(bad.Ok);
    }

    [Fact]
    public void RevokedDeviceRejected()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42");
        var roots = TrustedRoot(root);

        var before = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots);
        Assert.True(before.Ok, before.Reason);

        var after = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots, revoked: id => id == device.Did);
        Assert.False(after.Ok);
    }

    [Fact]
    public void DeviceRegistryTracksRevocation()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "charge", "api.bank", "https://api.bank/invoices");
        string action = SignDeviceAction(device, grant, "https://api.bank/invoices/42");
        var roots = TrustedRoot(root);

        var registry = new VouchFleet.DeviceRegistry();
        registry.Enroll(device.Did, grant);
        Assert.Single(registry.ActiveDevices());

        var before = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots, revoked: registry.IsRevoked);
        Assert.True(before.Ok, before.Reason);

        registry.Revoke(device.Did);
        Assert.Empty(registry.ActiveDevices());
        var after = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, roots, revoked: registry.IsRevoked);
        Assert.False(after.Ok);
    }

    [Fact]
    public void DidKeyLinkResolvesWithoutTrustMap()
    {
        VouchAgent root = VouchAgent.Create("root.example");
        VouchAgent device = VouchAgent.Create();
        string grant = VouchFleet.EnrollDevice(root, device.Did, "read", "t", "https://x/y");
        string action = SignDeviceAction(device, grant, "https://x/y/z");

        var result = VouchFleet.VerifyDelegatedChain(new[] { grant, action }, TrustedRoot(root));
        Assert.True(result.Ok, result.Reason);
    }
}
